src/lib.rs

```diff
@@ -70,298 +75,359 @@ pub trait UnicodeTruncateStr {
+            .rev()
+            // map to byte index and the width of char start at the index
+            .map(|(byte_index, char)| (byte_index, char.width().unwrap_or(0)))
             // fold to byte index and the width from end to the index
-            .scan(0, |w, (bidx, cw)| {
-                *w = *w + cw;
-                Some((bidx, *w))
+            .scan(0, |sum, (byte_index, char_width)| {
+                *sum += char_width;
+                Some((byte_index, *sum))
             })
-            .take_while(|&(_, w)| w <= width)
+            .take_while(|&(_, current_width)| current_width <= max_width)
             .last()
             .unwrap_or((self.len(), 0));
+        (self.get(byte_index..).unwrap(), new_width)
+    }
 
-        (self.get(bidx..).unwrap(), new_width)
+    #[allow(clippy::collapsible_else_if)]
+    #[inline]
+    fn unicode_truncate_centered(&self, max_width: usize) -> (&str, usize) {
+        if max_width == 0 {
+            return ("", 0);
+        }
+
+        let mut current_width: usize = self.width();
+        if current_width <= max_width {
+            return (self, current_width);
+        }
+
+        let mut iter = self
+            .char_indices()
+            // map to byte index and the width of char start at the index
+            .map(|(byte_index, char)| (byte_index, char.width().unwrap_or(0)))
+            // zero width doesn't need to be checked, they are always kept
+            .filter(|&(_, char_width)| char_width > 0);
+
+        let mut start_is_truncated = false;
+        let mut end_index = self.len();
+
+        // Amount of things taken from start / end. Tries to balance out to keep the center center.
+        let mut balance: isize = 0;
+
+        while current_width > max_width {
+            if balance >= 0 {
+                if let Some((byte_index, char_width)) = iter.next_back() {
+                    current_width = current_width.saturating_sub(char_width);
+                    end_index = byte_index;
+                    balance = balance.saturating_sub(char_width as isize);
+                } else {
+                    break;
+                }
+            } else {
+                if let Some((_, char_width)) = iter.next() {
+                    current_width = current_width.saturating_sub(char_width);
+                    start_is_truncated = true;
+                    balance = balance.saturating_add(char_width as isize);
+                } else {
+                    break;
+                }
+            }
+        }
+
+        // When truncation happened at the start then get the next byte_index as thats where it
+        // actually starts. Reason: index is where the char starts, not where it ends.
+        let start_index = if start_is_truncated {
+            iter.next().map_or(end_index, |(byte_index, _)| byte_index)
+        } else {
+            0
+        };
+
+        let result = self.get(start_index..end_index).unwrap();
```

Comment:
Let's add a comment here explaining that the `unwrap` is safe because `start_index` and `end_index` all come from in-range indices.

To me the extra unwrap means there are some constraints that are known to us human but not to the compiler. Wonder if there is an alternative way to organize the algorithm to remove the unwrap. That might lead to better performance because the compiler has more information to do optimization. But that can be done later.

Reply:
We could use get_unchecked which is unsafe. We can supply the SAFETY hint here.

Alternatively use `self[start_index..end_index]` which could panic. Also doesn't use an unwrap.

Reply:
IMHO the current way is good enough. This isn't on hot path so I feel it's not worth it to introduce unsafe. It's also better than `self[start_index..end_index]` because that's more explicit and will remind us if we do revisit this in the future :)